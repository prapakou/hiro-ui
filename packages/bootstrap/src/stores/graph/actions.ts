import { createAsyncAction } from "typesafe-actions";

import {
  GraphQueryRequest,
  GraphQuerySuccess,
  GraphQueryError
} from "./constants";

export const graphQuery = createAsyncAction(
  "GRAPH_QUERY_REQUEST",
  "GRAPH_QUERY_SUCCESS",
  "GRAPH_QUERY_FAILURE",
  "GRAPH_QUERY_CANCEL"
)<GraphQueryRequest, GraphQuerySuccess, GraphQueryError, { id: string }>();
