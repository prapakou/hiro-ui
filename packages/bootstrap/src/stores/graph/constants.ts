import { ActionType } from "typesafe-actions";
import { MappedTypes } from "@hiro-graph/orm-mappings";
import { GraphVertex, Entity } from "@hiro-graph/orm";

export const GRAPH_NAMESPACE = "graph";

export interface GraphQueryRequest {
  id: string;
  entity: MappedTypes;
  method: keyof Entity<GraphVertex>;
  args?: any;
}

export interface GraphQuerySuccess {
  id: string;
  response: any;
}

export interface GraphQueryError {
  id: string;
  error: string;
}

export type GraphQueryItem = {
  loading?: boolean;
} & Partial<GraphQueryRequest & GraphQuerySuccess & GraphQueryError>;

export interface GraphStateType {
  [index: string]: GraphQueryItem;
}
export type GraphActionsType = ActionType<typeof import("./actions")>;
