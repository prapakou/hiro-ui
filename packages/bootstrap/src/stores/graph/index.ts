export * from "./actions";
export * from "./constants";
export * from "./hooks";
export * from "./reducers";
export * from "./sagas";
