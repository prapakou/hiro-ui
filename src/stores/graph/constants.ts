import { ActionType } from "typesafe-actions";

export const GRAPH_NAMESPACE = "graph";

export interface TokenSet {
  token: string;
}

export type GraphStateType = Partial<TokenSet>;

export type GraphActionsType = ActionType<typeof import("./actions")>;
