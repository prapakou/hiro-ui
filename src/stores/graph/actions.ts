import { createStandardAction } from "typesafe-actions";

import { TokenSet } from "./constants";

export const tokenSet = createStandardAction("TOKEN_SET")<TokenSet>();
export const tokenClear = createStandardAction("TOKEN_CLEAR")();
