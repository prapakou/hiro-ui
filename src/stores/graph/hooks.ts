import { useCallback } from "react";
import { get } from "lodash-es";
import { useMappedState, useDispatch } from "redux-react-hook";

import { HIRO_NAMESPACE } from "../constants";

import { GRAPH_NAMESPACE, GraphStateType } from "./constants";
import { tokenSet, tokenClear } from "./actions";

export const useToken = () => {
  const mapState = useCallback(
    state => get(state, [HIRO_NAMESPACE, GRAPH_NAMESPACE]),
    []
  );

  const auth = useMappedState(mapState) as GraphStateType;

  return auth.token;
};

export const useTokenDispatch = () => {
  const dispatch = useDispatch();

  const setToken = useCallback(
    (token: string) => dispatch(tokenSet({ token })),
    [dispatch]
  );
  const clearToken = useCallback(() => dispatch(tokenClear()), [dispatch]);

  return { setToken, clearToken };
};
