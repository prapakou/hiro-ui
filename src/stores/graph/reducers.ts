import { createReducer } from "typesafe-actions";
import { omit } from "lodash-es";

import { GraphStateType, GraphActionsType } from "./constants";

export const graphReducer = createReducer<GraphStateType, GraphActionsType>({})
  .handleAction("TOKEN_SET", (state, { payload }) => ({
    ...state,
    token: payload.token
  }))
  .handleAction("TOKEN_CLEAR", state => ({ ...omit(state, ["token"]) }));
