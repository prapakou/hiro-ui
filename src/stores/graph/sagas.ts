import { takeEvery, take, call, put, race, delay } from "redux-saga/effects";
import { ActionType } from "typesafe-actions";

import { Orm } from "../../contexts";

import { GraphQueryRequest } from "./constants";
import { graphQuery } from "./actions";

function* sendRequest(req: any, id: string, params: any[]) {
  try {
    const response = yield call(req, ...params);
    yield put(graphQuery.success({ id, response }));
  } catch (error) {
    yield put(graphQuery.failure({ id, error }));
  }
}

function* handleRequest(
  orm: Orm,
  action: ActionType<typeof graphQuery.request>
) {
  const { id, entity, method, args } = action.payload as GraphQueryRequest;
  const params = Array.isArray(args) ? args : [args];

  yield race({
    result: yield call(sendRequest, orm[entity][method], id, params),
    cancel: yield take(graphQuery.cancel)
  });
}

export function* handleGraphRequests(orm?: Orm) {
  if (!orm) {
    console.warn(
      `HiroGraph: No orm available will not run saga handleGraphRequests`
    );
    return;
  }

  // @ts-ignore
  yield takeEvery(graphQuery.request, handleRequest, orm);
}
