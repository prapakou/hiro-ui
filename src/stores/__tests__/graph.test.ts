import configureStore, { MockStoreEnhanced } from "redux-mock-store";

import { tokenSet, tokenClear, graphReducer } from "../graph";

const mockStore = configureStore();
let store: MockStoreEnhanced;

beforeEach(() => {
  store = mockStore({});
});

test("Auth state", () => {
  let action: any = store.dispatch(tokenSet({ token: "My token" }));
  expect(store.getActions()).toMatchSnapshot("Set");
  expect(graphReducer(undefined, action)).toMatchSnapshot("Add - Reducer");

  action = store.dispatch(tokenClear());
  expect(store.getActions()).toMatchSnapshot("Clear");
  expect(graphReducer(undefined, action)).toMatchSnapshot("Clear - Reducer");
});
