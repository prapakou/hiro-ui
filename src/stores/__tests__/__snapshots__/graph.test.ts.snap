// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth state: Add - Reducer 1`] = `
Object {
  "token": "My token",
}
`;

exports[`Auth state: Clear - Reducer 1`] = `Object {}`;

exports[`Auth state: Clear 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": Object {
      "token": "My token",
    },
    "type": "TOKEN_SET",
  },
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "TOKEN_CLEAR",
  },
]
`;

exports[`Auth state: Set 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": Object {
      "token": "My token",
    },
    "type": "TOKEN_SET",
  },
]
`;
